### ISSUE TITLE
Trouble customizing markers

### ISSUE THREAD

**Issue by jodonnell** [bbc/peaks.js#286]
Hello, I'm having trouble customizing the markers.  It seems the documentation might be out of date?  The listed functions like segmentInMarker, don't seem to exist anymore.  I viewed the source and I believe I want to be using createPointMarker.  When I edit the source it looks good.

However when i pass a different function into my options it seems like it gets written on top of.

```
        Utils.extend(this.options, opts);
        console.log(this.options)
        Utils.extend(this.options, {
            createSegmentMarker: mixins.createSegmentMarker,
            createSegmentLabel: mixins.createSegmentLabel,
            createPointMarker: mixins.createPointMarker
        });
```
shows this.options having the original copy inside of it rather than the one I passed in.  Even if that passed, it appears the next line merges the defaults on top of what i would pass in options.

How should I be getting my custom createPointMarker function to be used?

Thanks again!!

**Comment by jodonnell** [comment]
This seems to work, is it advisable?
```
        Peaks.init(options, (err, peaks) => {
            if (err)
                return;
            peaks.options.createPointMarker = createPointMarker;
            peaks.points.add({time: 0.1, editable: true, color: 'ff0000', labelText: 'Hot Damn'});
```

**Comment by jodonnell** [comment]
Hmm, also is it possible to have the marker render differently (have a different function for each?) in the zoom view and overview view?

**Comment by chrisn** [comment]
I think we need to revisit the API design for marker customisation, I don't know if anyone is using the existing documented API, and as you say, the code and documentation have got out of sync. What kinds of customisations do you need? Do you need full control to create your own shapes, or would simpler customisations such as marker size and colour be enough?

**Comment by jodonnell** [comment]
here is what we want it to look like
![Screen Shot 2019-11-05 at 2 00 11 PM](https://user-images.githubusercontent.com/217747/68237161-9c2f3580-ffd4-11e9-83ff-8a862836ef35.png)

i've already implemented this
![Screen Shot 2019-11-05 at 2 00 45 PM](https://user-images.githubusercontent.com/217747/68237238-c254d580-ffd4-11e9-956d-2298a948d41c.png)

so i'm implementing the bottom part just fine, but its reflecting in the top part, when in the top part( the overview) i need it to be different.  i think i could do something hacky like in the createPointMarker look at the group's parent parent parent container, check if its overview or zoom and render different things completely.

so definitely on the more custom side.

**Comment by chrisn** [comment]
I just pushed a fix for the `createPointMarker` etc options, so these should now be working.

**Comment by jodonnell** [comment]
Hi Chris, that worked.  I changed my dependency to "peaks.js": "git+ssh://git@github.com/bbc/peaks.js.git",
then had to go into node_modules/peak.js
npm i
then 
`npx browserify -d -e ./src/main.js -t deamdify -s peaks | npx exorcist peaks.js.map | npx derequire - > peaks.js`

so unless there is a better way we would have to wait for you to create a release.

but thats no big deal cause `peaks.options.createPointMarker = createPointMarker;` works for now.

Thank you!

**Comment by chrisn** [comment]
Hi Jacob,

Not sure there's a better way, to be honest. Does `npm run build` not work within `node_modules/peaks.js`? To make things easier for you, I'll make a new bugfix release in a moment...

To allow per-view customisation, I'm thinking we should extend the [view settings API](https://github.com/bbc/peaks.js#view-settings-api) rather than add more config options to `Peaks.init()`.

```javascript
const createPointMarker = (options) => {
  // Returns a Konva.Group object
};

Peaks.init(options, (err, peaks) => {
  const overview = peaks.views.getView('overview');
  overview.setPointMarkerFactory(createPointMarker);
});
```

I'd like to revisit the options being passed to `createPointMarker`, and the return type - perhaps we can pass in an empty `Konva.Group` for the function to populate? And could we move all the event handler stuff out of `createPointMarker`, as I assume this wouldn't generally need customising? I'd need to do some experiments to see what works.

What do you think?

**Comment by jodonnell** [comment]
no, running those commands tried to kick off npm run lint which failed (didn't investigate why running run build would kick off lint...)

setting them per view sounds good to me... although would it do a render after init, but before setPointMarkerFactory has run?  Like it would flash the old marker and then show the new?  Otherwise seems good.

Hmm, i do need options.height to draw a line,
```
    const line = new Konva.Line({
        x: 0,
        y: 0,
        points: [0, 0, 0, options.height],
        stroke: '#6610F2',
        strokeWidth: 1
    });
```

as far as the handlers i am customizing that slightly,
```
    handle.on('mouseover touchstart', () => {
        options.layer.draw();
        document.getElementById('zoomview-container').style.cursor = 'grab';
    });

    handle.on('mouseout touchend', () => {
        options.layer.draw();
        document.getElementById('zoomview-container').style.cursor = 'default';
    });

    group.on('dragstart', () => {
        options.layer.draw();
        document.getElementById('zoomview-container').style.cursor = 'grabbing';
    });

    group.on('dragend', () => {
        options.layer.draw();
        document.getElementById('zoomview-container').style.cursor = 'default';
    });
```
so the none defaults there are changing the mouse.

also if you look at the design, the top purple with the dots is the only part that you can grip, so i have the grip events on the handle not the group.

maybe passing an empty group would make sense, but if the drag stuff was left in place you would need to update the group to take
``` 
        draggable: options.draggable,
        dragBoundFunc
```
so at the end of the day, as a customizer they feel equivalent to me

**Comment by chrisn** [comment]
> setting them per view sounds good to me... although would it do a render after init, but before setPointMarkerFactory has run? Like it would flash the old marker and then show the new? Otherwise seems good.

I'm thinking we'd disallow passing points and segments to `Peaks.init()`, so the application could set up the factories before rendering.

Thanks for the info on options, it's helpful to have a concrete example.

**Comment by jodonnell** [comment]
sounds good to me, it would be an api change so any consumers would need to update probably a very small amount of code, but if you're good with that works for me!

**Comment by chrisn** [comment]
I had a quick look into this. To support view resizing, we need to be able to also resize the markers. So I think we need to define an API that allows Peaks.js to request markers to resize themselves based on the new view container size. 

At the moment, `createPointMarker` and `createSegmentMarker` return a `Konva.Group`, but instead we need to return a higher level interface that can (at least) receive resize requests.

**Comment by chrisn** [comment]
@jodonnell I made a new [marker-customisation](https://github.com/bbc/peaks.js/tree/marker-customisation) branch with a new API for creating custom point/segment markers. There's [documentation](https://github.com/bbc/peaks.js/blob/marker-customisation/customising.md) and [an example](https://github.com/bbc/peaks.js/blob/marker-customisation/demo/custom-markers.html). This is a breaking change from the current API, so I'd like to hear your feedback before merging. Thanks!

**Comment by jodonnell** [comment]
I think I will have time to try and integrate it into our project tomorrow, I'll let you know how it goes.

**Comment by chrisn** [comment]
Thank you! I'm going to take another look at the API to see if there's anything that could be simplified.

**Comment by jodonnell** [comment]
Hi Chris, one problem I'm having is I need to pass data into these functions.  Specifically, i kind of have two copies of the points, one lives inside peaks, but the other lives inside react.  When we interact with the peaks markers we need the react points to be updated as well so the UI updates.

The problem here is we have a code path like this...

```
export const createPointMarkerWrapper = function createPointMarkerWrapper(
    points,
    setPoints,
    removePoint
) {
    const createPointMarker = function createPointMarker(options) {
        const point = find(points, { id: options.point.id });
        if (options.view === 'overview')
            return new OverviewPointMarker(options, point);
        return new ZoomPointMarker(options, point, points, setPoints, removePoint);
    };
    return createPointMarker;
};

class OverviewPointMarker extends Peaks.PointMarker {
    constructor(options, point) {
        super(options);
        this.point = point;
    }

    createMarker(group, options) {
        overviewGroup(group, options, this.point);
    }
.....
}
```

the problem here is the constructor calls createMarker at which point this.point hasn't been defined yet.

i suppose i can shove what i need in the options object, that feels a bit funny

**Comment by jodonnell** [comment]
Another question, the method bindEventHandlers, why does that exist?  As I was refactoring to using this branch i left the handlers in the createMarker method and it seems to work fine.

It seems otherwise I need to put the konva objects into an instance variable so I can access in bindEventHandlers...

**Comment by jodonnell** [comment]
fitToView, really i would just want createMarker to be called again where i use options.layer.getHeight().

do i have to instead save off any objects that have height changed and modify them?

i understand if I have to for some technical limitations, but more code, more bugs!

**Comment by jodonnell** [comment]
and positionUpdated is just there if you need to hook into that, correct?

**Comment by jodonnell** [comment]
the one thing thats not working is something I never really understood, at the bottom of the handle we display the time

```
    const time = new Konva.Text({
        x: handleX + 10,
        y: timeSectionStartY + 5,
        text: '',
        fontSize: 10,
        fontFamily: 'Rubik',
        fill: '#fff',
        textAlign: 'center',
        opacity: 0.8
    });

    // eslint-disable-next-line no-param-reassign
    group.time = time;
```
thats no longer working, i never understood why assigning group.time actually set the time, but it did somehow, and now it no longer does,

although this seems to work
```
    const time = new Konva.Text({
        x: handleX + 10,
        y: timeSectionStartY + 5,
        text: options.point.time,
        fontSize: 10,
        fontFamily: 'Rubik',
        fill: '#fff',
        textAlign: 'center',
        opacity: 0.8
    });
```
so thats good

**Comment by jodonnell** [comment]
generally i would say the api is a little more complicated now and i think the additional hook points to solve any problems for me, but probably nice to have the additional flexibility

the biggest problem this solves for us is it gets rid of the hacky way we were determining if it was the overview or zoom view.

i think the thing that would improve our code the most is we need the create markers code to have access to some of our variables and the path to that right now is very complicated...

```
    const createPointMarker = createPointMarkerWrapper(points, setPoints, removePoint);

    React.useEffect(() => {
        const options = {
            height: 169,
            zoomWaveformColor: '#a8a0a066',
            overviewWaveformColor: '#a8a0a04c',
            pointMarkerColor: '#dd4a5f',
            overviewHighlightColor: '#4d586d',
            overviewHighlightOffset: 0,
            axisGridlineColor: '#aaabaf',
            axisLabelColor: '#21343a',
            zoomLevels: [512, 1024, 2048, 4096, 8192],
            zoomAdapter: 'static',
            containers: {
                overview: document.getElementById('overview-container'),
                zoomview: document.getElementById('zoomview-container')
            },
            mediaElement: document.querySelector('audio'),
            dataUri: {
                arraybuffer: peaksUrl,
            },
            createPointMarker
        };

        Peaks.init(options, (err, peaks) => {
            if (err)
                return;

            peaks.zoom.setZoom(2);
            peaksRef.current = peaks;

            addPointsToPeaks();
            setIsLoaded(true);
        });
    }, []);

    React.useEffect(() => {
        // This could be trouble, like if points are loaded from server
        // before peaks is ready
        if (!peaksRef.current)
            return;
        // This updates the closure around peaks internal createPointMarker
        // so that when points changes the function has the update value
        peaksRef.current.options.createPointMarker = createPointMarker;

        peaksRef.current.points.removeAll();
        addPointsToPeaks();
    }, [points]);
```

to keep the variables up to date in the createPointMarker i need to reassign to options.createPointMarker, whenever points changes

**Comment by chrisn** [comment]
Thanks for all your feedback, I'll read through in more detail later. I think I want to change the API to remove the object inheritance, which should simplify things a little.

**Comment by jodonnell** [comment]
👍

**Comment by chrisn** [comment]
I pushed an update earlier which removes the inheritance. 

> bindEventHandlers, why does that exist?

I added this because the marker object would generally need to do two things: create the Konva objects, then add event handlers. So I guessed it would be useful put these into separate functions. I've removed this in the latest update, so you can now do everything from the marker object's `init()` function.

Initialisation of the marker object now happens in two steps: the constructor receives the `options` passed to `createPointMarker`, then `init()` receives the `Konva.Group`.

> i never understood why assigning group.time actually set the time

There's code in `PointsLayer` that updates the time (e.g., when markers are dragged) [here](https://github.com/bbc/peaks.js/blob/master/src/main/views/points-layer.js#L285).

I changed the object structure so that we're not adding fields to the Konva objects - TypeScript doesn't like this, and I would like (eventually) to migrate the code to TS.

> and positionUpdated is just there if you need to hook into that, correct?

Yes, you wouldn't implement `positionUpdated` unless you need to do something when the marker is moved. In the `DefaultPointMarker` object, this is where we now update the `time` value. The reason for this change was to move all the marker behaviour out of `PointsLayer`, which shouldn't know about the details of how the markers are rendered, it just manages creating and removing them.

> i think the thing that would improve our code the most is we need the create markers code to have access to some of our variables

Would it help to be able to attach your own data to the `Point`? For example,

```javascript
const point = peaks.points.add({
  time: time,
  labelText: 'Test point',
  editable: true,
  userData: <any value you want>
});
```

**Comment by jodonnell** [comment]
Hi Chris, I updated our code to the latest example.  I think getting rid of the inheritance was probably a positive move.

I'm trying to think of the answer to your question but its pretty complicated so I'm not sure I could really tell without trying to work through it.

The gist is we are passing an array of points with some of our data and a function to set that array of points.  When we call the function to set the points it creates a copy (react requires this) so we then need to somehow update peaks to get the new array of points (which you won't get until the next render).

So I *think* with what you have above we would need still need to change this userData on every change to setPoints which I don't think would make this any easier.

This maybe just some necessary complexity of integrating with react.

I suppose one possibility is if instead of having the createPointMarker function, if we directly set the two classes, that would probably make the code more straight forward for us.


it would make the code i posted above look more like this

```
    React.useEffect(() => {
        const options = {
            height: 169,
            zoomWaveformColor: '#a8a0a066',
            overviewWaveformColor: '#a8a0a04c',
            pointMarkerColor: '#dd4a5f',
            overviewHighlightColor: '#4d586d',
            overviewHighlightOffset: 0,
            axisGridlineColor: '#aaabaf',
            axisLabelColor: '#21343a',
            zoomLevels: [512, 1024, 2048, 4096, 8192],
            zoomAdapter: 'static',
            containers: {
                overview: document.getElementById('overview-container'),
                zoomview: document.getElementById('zoomview-container')
            },
            mediaElement: document.querySelector('audio'),
            dataUri: {
                arraybuffer: peaksUrl,
            },
            createPointMarker
        };

        Peaks.init(options, (err, peaks) => {
            if (err)
                return;

            peaks.zoom.setZoom(2);
            peaksRef.current = peaks;

            addPointsToPeaks();
            setIsLoaded(true);
        });
    }, []);

    React.useEffect(() => {
        // This could be trouble, like if points are loaded from server
        // before peaks is ready
        if (!peaksRef.current)
            return;
        // This updates the closure around peaks internal createPointMarker
        // so that when points changes the function has the update value

        peaksRef.current.options.overviewPointMarker = new OverviewPointMarker(options, point);
        peaksRef.current.options.zoomPointMarker = new ZoomPointMarker(options, point, points, setPoints, removePoint);
        peaksRef.current.points.removeAll();
        addPointsToPeaks();
    }, [points]);
```

and remove the complicated function that creates a closure createPointMarkerWrapper
```
export const createPointMarkerWrapper = function createPointMarkerWrapper(
    points,
    setPoints,
    removePoint
) {
    const createPointMarker = function createPointMarker(options) {
        const point = find(points, { id: options.point.id });
        if (options.view === 'overview')
            return new OverviewPointMarker(options, point);
        return new ZoomPointMarker(options, point, points, setPoints, removePoint);
    };
    return createPointMarker;
};

```

**Comment by jodonnell** [comment]
I think the change I posted is probably a little confusing so to restate in a different way....
currently to get our data into the peaks render functions we have put the createPointMarker in a closure that captures the variables we need to send in.  When one of those variables changes 
```
React.useEffect(() => {
        // this is called when the variable is changed if you are not familiar with react
    }, [points]);```
we reset the createPointMarker function with the new closure around the updated variables.


with the idea i proposed above, with directly setting the class we don't need to do any of that, we can just update the classes with a new instantiation on change.

**Comment by chrisn** [comment]
A quick progress update: I have changed the API a little, replacing `positionUpdated` with `timeUpdated`, which is only called when the marker's timestamp is changed.

I'm having some issues getting the TypeScript definitions to work though, so PR #300 isn't yet ready to merge.

**Comment by jodonnell** [comment]
Okay great, thank you Chris.

**Comment by adamteale** [comment]
hi @chrisn are these changes available in v0.17.0? thanks!

**Comment by chrisn** [comment]
This is still work in progress, on the [`marker-customisation`](https://github.com/bbc/peaks.js/tree/marker-customisation) branch.

**Comment by adamteale** [comment]
Great, thank you @chrisn.

**Comment by chrisn** [comment]
These changes are now merged, and published in v0.18.0. @jodonnell is there anything more you need for marker customisation, or can we close this issue?

**Comment by jodonnell** [comment]
oh dang didn't realize i never replied.  we just merged this in and it seems to be working great.

Thank you Chris!
